src/cmd/services/m3coordinator/ingest/carbon/ingest.go

```diff
@@ -148,9 +149,107 @@ type ingester struct {
 	metrics              carbonIngesterMetrics
 	tagOpts              models.TagOptions
 
+	lineResourcesPool pool.ObjectPool
+
+	sync.RWMutex
 	rules []ruleAndRegex
+}
 
-	lineResourcesPool pool.ObjectPool
+func (i *ingester) OnUpdate(clusterNamespaces m3.ClusterNamespaces) {
+	rules := i.regenerateIngestionRules(clusterNamespaces)
+	if rules == nil {
+		namespaces := make([]string, 0, len(clusterNamespaces))
+		for _, ns := range clusterNamespaces {
+			namespaces = append(namespaces, ns.NamespaceID().String())
+		}
+		i.logger.Warn("generated empty carbon ingestion rules from latest cluster namespaces update. leaving"+
+			" current set of rules as-is.", zap.Strings("namespaces", namespaces))
+		return
+	}
+
+	compiledRules, err := i.compileRules(*rules)
+	if err != nil {
+		i.logger.Error("failed to compile latest rules. continuing to use existing carbon ingestion "+
+			"rules", zap.Error(err))
+		return
+	}
+
+	i.Lock()
+	i.rules = compiledRules
+	i.Unlock()
+}
+
+func (i *ingester) regenerateIngestionRules(clusterNamespaces m3.ClusterNamespaces) *CarbonIngesterRules {
+	var (
+		rules = &CarbonIngesterRules{
+			Rules: i.opts.IngesterConfig.RulesOrDefault(clusterNamespaces),
+		}
+		namespacesByRetention = make(map[m3.RetentionResolution]m3.ClusterNamespace, len(clusterNamespaces))
```

Comment:
So this is likely to be a very short list; may be worth considering making this into a slice and then check for overwrites when looping over this; although it's a very cold\ path, good habit to get into :+1:

Reply:
Not sure I follow. We're using this just to ensure that the storage policies on carbon rules map to namespaces that currently exist. Wouldn't converting this into a slice be less efficient here?